### ISSUE TITLE
Error running BatchAdjust

### ISSUE THREAD

**Issue by sanhe374** [CUHIMSR/CytofBatchAdjust#6]
Hi,

I get an error when I run the following:
BatchAdjust(
  basedir="P:/LiU/Postdok/Projekt/GraMS/Resultat/Analysis/CyTOF/3_BatchAdjustedFiles/",
  outdir="P:/LiU/Postdok/Projekt/GraMS/Resultat/Analysis/CyTOF/3_BatchAdjustedFiles/outdir/",
  channelsFile = "ChannelsToAdjust.txt",
  batchKeyword="Batch_",
  anchorKeyword = "anchor",
  method="95p",
  transformation=FALSE,
  addExt=NULL,
  plotDiagnostics=TRUE)

Error: Error in fcsTextParse(txt, emptyValue = emptyValue) : 
  Empty keyword name detected!If it is due to the double delimiters in keyword value, please set emptyValue to FALSE and try again!

**Comment by SamGG** [comment]
Hi,
This error is raised by flowCore package when reading one of the FCS files.
Do you think you could correct it on your own?
Best.

**Comment by sanhe374** [comment]
No, because I don't understand what the problem is. I have pre-gated the files in Cytobank and exported them as FCS files. When I try using these files I get the error but if I use fcs files that have not been exported from Cytobank there is no problem except that the results look at bit weird.

**Comment by SamGG** [comment]
If I got one FCS I can investigate and explain you. But I don't have time yet.
I pushed a patch that allows you to change the emptyValue flag from TRUE (default) to FALSE.
a) use my fork of this code at https://github.com/i-cyto/CytofBatchAdjust
b) before executing the adjustment, run `options(read.FCS.emptyValue = FALSE); options("read.FCS.emptyValue")` which will change the value of emptyValue argument in all the calls to read.FCS in the cytofBatchAdjust framework.
I didn't test it yet. Let me know if it works for you.
PS: I think you can close your previous issue.

**Comment by sanhe374** [comment]
Thank you, I did not manage to get it working.
Error in fcsTextParse(txt, emptyValue = emptyValue) : 
  Not compatible with requested type: [type=list; target=logical]

I am appending one of the files if you have time to take a look at it. 

[3_BatchAdjustedFiles.zip](https://github.com/CUHIMSR/CytofBatchAdjust/files/5651268/3_BatchAdjustedFiles.zip)

**Comment by SamGG** [comment]
What is your R and flowCore version?
`sapply(c("base", "flowCore"), packageVersion)`

**Comment by sanhe374** [comment]
$base
[1] 4 0 2

$flowCore
[1] 2 0 1

**Comment by SamGG** [comment]
I am using R 4.0.3 and flowCore 2.2 (lastest stable bioconductor, not the dev). I simply loaded the FCS file of the zip file and didn't get any warning. Could you try this with your current setup?
``` r
library(flowCore)
ff = read.FCS("C:/demo/201207-sahne/c01_06122018_01_0_0_Batch_1_MS12.0_comped.fcs.fcs", transformation = FALSE)
ff
#> flowFrame object 'c01_06122018_01_0_0_Batch_1_MS12.0_comped.fcs'
#> with 94634 cells and 69 observables:
#>             name        desc     range  minRange  maxRange
#> $P1         Time          NA  33554432         0  33554431
#> $P2       length          NA       128         0       127
#> $P3       Center          NA     16384         0     16383
#> $P4       Offset          NA      1024         0      1023
#> $P5        Width          NA       512         0       511
#> ...          ...         ...       ...       ...       ...
#> $P65     Pt194Di       194Pt     16384         0     16383
#> $P66     Pt195Di       195Pt     32768         0     32767
#> $P67     Pt198Di         CD8     65536         0     65535
#> $P68     Pb208Di       208Pb      1024         0      1023
#> $P69     Bi209Di        CD16      4096         0      4095
#> 497 keywords are stored in the 'description' slot
```

<sup>Created on 2020-12-07 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by SamGG** [comment]
> Error in fcsTextParse(txt, emptyValue = emptyValue) :
Not compatible with requested type: [type=list; target=logical]

I have just pushed a new release that corrects my error. I think it should be OK now although still untested.

**Comment by sanhe374** [comment]
Is there something I need to do to get access to the new release?

**Comment by SamGG** [comment]
Just reload from github as you did previously.
Did you replicate the 3 lines of code and get an error with your current setup?

**Comment by sanhe374** [comment]
![PrePostVariance](https://user-images.githubusercontent.com/48206476/101340192-ae3a7200-387f-11eb-9f15-452c12b5d972.png)
Thank you 👍 I do not get an error when running the 3 lines of code and now the whole script seems to work without problem. I just have ask also about the output (see enclosed figure). In many instances it looks like the post-variance is higher than the pre-variance. Is that common or correct?

**Comment by SamGG** [comment]
If you didn't get an error when running the 3 lines of code and your initial setup, I am perplex about the initial error concerning emptyValue.

Concerning the output, I don't have any clear understanding of it. My colleague and I prefer to inspect the density plot of the output FCS files that are the real data you are going to work with. A quick view would be to compute a tSNE reduction of dimensions of sampled cells of the anchor and to color them according their batch origin. If the batch effect is removed, colors should overlap.

In our experience, the percent defined in the method depends on the channel: for some channels, 90p is OK, for others with a small positive peak, 99p is needed to notice this peak. When relying on a lower percentile, the positive peak might be corrected from an effect that is observed in the negative peak only (eg 6th from the bottom). In the following figure, 95p would be a very bad choice, and no correction is probably better than anything else.

If you are in Europe, contact me directly. Otherwise I think you have to build your own experience with this useful tool.

![image](https://user-images.githubusercontent.com/3333348/101347643-af24d100-388a-11eb-9c71-e45bfc3106ba.png)
